docs/language/informal/toplevel-inference.md

```diff
@@ -0,0 +1,311 @@
+
+Status: Proposal
+
+
+## Top level inference overview
+
+Top level inference has two phases:
+
+1. **Method override inference**
+    * If you omit a return type or parameter type from an overridden method,
+    inference will try to fill in the missing type using the signature of the
+    method you are overriding.
+2. **Static variable and field inference** 
+    * If you omit the type of a field, setter, or getter, which overrides a
+   corresponding member of a superclass, then inference will try to fill in the
+   missing type using the type of the corresponding member of the superclass.
+    * Otherwise, declarations of static variables and fields that omit a type
+   will be inferred from their initializer if present.
+
+All results from the first phase are available in the second phase.
+
+
+As a general principle, when inference fails it is an error.  Tools are free to
+behave in implementation specific ways to provide a graceful user experience,
+but with respect to the language and its semantics, programs for which inference
+fails are unspecified.
+
+
+The intuitive idea behind this proposal is that top level inference works by
+first inferring methods, and then doing inference for "field like things"
+(static variables, fields, setters, and getters), where we don't allow inference
+for "field like things" to depend on the results of inference for other "field
+like things" except in cases where the dependencies are statically predictable.
+
+
+Some broad principles for type inference that the language team has agreed to,
+and which this proposal is designed to satisfy:
+* Type inference should be free to fail with an error, rather than being
+  required to always produce some answer
+* It should not be possible for a programmer to observe a declaration as having
+  two different types at difference points in the program (e.g. dynamic for
+  recursive uses, but subsequently int).  Some consistent answer (or an error)
+  should always be produced.  See the example below for an approach that
+  violates this principle.
+* The inference for local variables, toplevel variables, and fields, should
+  either agree or error out.  The same expression should not be inferred
+  differently at different syntactic positions.  It’s ok for an expression to be
+  inferrable at one level but not at another.
+* Obvious types should be inferred
+* Inferred and annotated types should be treated the same
+* Simple intuition for users
+* Efficient to implement
+* Type arguments should be mostly inferred
+
+
+A motivating example for the current strong mode implementation (that also
+serves as a demonstration of its shortcomings) is as follows:
+
+
+```dart
+class A {
+  var x = 3;
+}
+var a = new A();
+var b = a.x;
+```
+
+
+To ensure stability with respect to cycles and different orderings, current
+strong mode does inference in two passes: first on the static variables, and
+secondly on fields (method override based inference is mostly irrelevant for
```

Comment:
It makes sense to mention top-level variables because they are in the example: 'first[ly?] on the static [and top-level] variables, and', or maybe 'secondly' --> 'then'.